charms/tensorboard-controller/src/charm.py

```diff
@@ -1,184 +1,302 @@
+from ops.pebble import CheckStatus, Layer
 
+PROBE_PORT = "8081"
+PROBE_PATH = "/healthz"
 
-class CheckFailed(Exception):
-    """Raise this exception if one of the checks in main fails."""
+K8S_RESOURCE_FILES = [
+    "src/templates/auth_manifests.yaml.j2",
+]
+CRD_RESOURCE_FILES = [
+    "src/templates/crds.yaml.j2",
+]
 
-    def __init__(self, msg, status_type=None):
-        super().__init__()
 
-        self.msg = str(msg)
-        self.status_type = status_type
-        self.status = status_type(self.msg)
+class TensorboardController(CharmBase):
+    """Tensorboard Controller Charmed Operator."""
 
-
-class Operator(CharmBase):
     def __init__(self, *args):
         super().__init__(*args)
 
-        self.log = logging.getLogger(__name__)
-        self.image = OCIImageResource(self, "oci-image")
+        self.logger = logging.getLogger(__name__)
+
+        # retrieve configuration and base settings
+        self._namespace = self.model.name
+        self._lightkube_field_manager = "lightkube"
+        self._name = self.model.app.name
+        self._exec_command = "/manager"
+        self._container_name = "tensorboard-controller"
+        self._container = self.unit.get_container(self._container_name)
+
+        # setup context to be used for updating K8s resources
+        self._context = {
+            "app_name": self._name,
+            "namespace": self._namespace,
+            "service": self._name,
+        }
+        self._k8s_resource_handler = None
+        self._crd_resource_handler = None
+
         self.gateway = GatewayRequirer(self)
 
-        for event in [
-            self.on.install,
-            self.on.leader_elected,
-            self.on.upgrade_charm,
-            self.on.config_changed,
-            self.on["gateway-info"].relation_changed,
-        ]:
-            self.framework.observe(event, self.main)
+        # setup events to be handled by main event handler
+        self.framework.observe(self.on.config_changed, self._on_event)
+        self.framework.observe(self.on.leader_elected, self._on_event)
+        self.framework.observe(self.on.tensorboard_controller_pebble_ready, self._on_event)
+        self.framework.observe(self.on["gateway-info"].relation_changed, self._on_event)
 
-    def main(self, event):
-        try:
-            self._check_leader()
+        # setup events to be handled by specific event handlers
+        self.framework.observe(self.on.install, self._on_install)
+        self.framework.observe(self.on.upgrade_charm, self._on_upgrade)
+        self.framework.observe(self.on.remove, self._on_remove)
+        self.framework.observe(self.on.update_status, self._on_update_status)
 
-            image_details = self._check_image_details()
-        except CheckFailed as check_failed:
-            self.model.unit.status = check_failed.status
-            return
+    @property
+    def container(self) -> Container:
+        """Return container."""
```

Comment:
I think I'll only keep the property and have it fetch the container object every time it's called.

Reply:
> I think I'll only keep the property and have it fetch the container object every time it's called.

I won't touch this for now. I believe it's main purpose is to establish a differentiation between read-only and read-write properties. The rationale is that we always maintain a managed attribute in a field like `self._container`, which we can then expose using a "public" container property. Then if we also want to make this attribute writeable, we can also define a setter. Since Python doesn't really support private fields, these are all mostly semantics. The important thing is that we're consistent (I see we already use this pattern in chisme, and the operator framework, let's stick to it).